src/sentry/tagstore/snuba/backend.py

```diff
@@ -157,21 +162,50 @@ def __get_tag_keys_for_projects(
         limit=1000,
         keys=None,
         include_values_seen=True,
+        use_cache=True,
         **kwargs
     ):
-        filters = {"project_id": projects}
+        filters = {"project_id": sorted(projects)}
         if environments:
-            filters["environment"] = environments
+            filters["environment"] = sorted(environments)
         if group_id is not None:
             filters["group_id"] = [group_id]
         if keys is not None:
-            filters["tags_key"] = keys
+            filters["tags_key"] = sorted(keys)
         aggregations = [["count()", "", "count"]]
 
         if include_values_seen:
             aggregations.append(["uniq", "tags_value", "values_seen"])
         conditions = []
 
+        should_cache = (
+            random.random() <= options.get("snuba.tagstore.cache-tagkeys-rate")
+            and use_cache
+            and group_id is None
+        )
+
+        if should_cache:
+            filtering_strings = [
+                u"{}={}".format(key, value) for key, value in six.iteritems(filters)
+            ]
+            cache_key = u"testing.tagstore.__get_tag_keys:{}".format(
+                md5_text(*filtering_strings).hexdigest()
+            )
+            # Round times by 5 minutes since we cache for 5 minutes
+            if start:
+                start = start.replace(minute=start.minute // 5 * 5, second=0, microsecond=0)
+                cache_key += ":" + start.isoformat()
+            if end:
+                end = end.replace(minute=end.minute // 5 * 5, second=0, microsecond=0)
+                cache_key += ":" + end.isoformat()
```

Comment:
This is likely going to cause a thundering herd every five minutes creating a load spike on snuba, since everyone will refresh the cache at the same time. Could you please offset the keys for different projects ?

Reply:
Another thing I thought about. Could you investigate whether we have use cases where the datetime is not actually passed at all ? I suspect that may be the largest use case looking at the queries I see in snuba (they mostly query tag keys starting exactly 3 months before the query is done and end one second before the query is done - which is the default timeframe added by snuba when it is missing). If this is called without timeframe, then shifting the datetime to avoid spikes won't work, we will have to add jitter to the TTL parameter in the call to memcache

Reply:
:\ Dang you're right it definitely isn't being passed. I think I'll open a separate PR to add the timeframe like we did in #16176 first instead of doing the TTL jitter.

Reply:
Wait, that's a different problem. If they are not passed, that is good for this goal, it means that the user wants "all the tags for the organization", this means we can be even more aggressive on caching. You have a simpler cache key and you can use the TTL for jitter. I think you should support a reasonable cache policy in both cases and then, as a followup optimization, reduce the scope of the queries to what the user needs .

Reply:
So the idea would be to create `cache_keys` without start/end time since we know the endpoint won't be passing anything except a 3 month period. Then in a later optimization introduce the start/end times back in?

As well I added a 60s jitter to the cache TTL, which is hopefully enough to avoid a thundering herd.